modules/mtl/arrow-mtl/src/main/kotlin/arrow/mtl/extensions/accumt.kt

```diff
@@ -0,0 +1,193 @@
+}
+
+@extension
+interface AccumtTMonadTrans<S> : MonadTrans<Kind<ForAccumT, S>> {
+
+  fun MS(): Monoid<S>
+
+  override fun <G, A> Kind<G, A>.liftT(MG: Monad<G>): Kind2<Kind<ForAccumT, S>, G, A> =
+    AccumT(MG) { _: S ->
+      MG.run {
+        flatMap { a ->
+          MG.just(MS().empty() toT a)
+        }
+      }
+    }
+}
+
+@extension
+interface AccumTAlternative<S, F> : Alternative<AccumTPartialOf<S, F>>, AccumTApplicative<S, F> {
+
+  fun AF(): Alternative<F>
+  override fun MF(): Monad<F>
+  override fun MS(): Monoid<S>
+
+  override fun <A> Kind<AccumTPartialOf<S, F>, A>.orElse(b: Kind<AccumTPartialOf<S, F>, A>): Kind<AccumTPartialOf<S, F>, A> =
+    (this.fix() to b.fix()).let { (ls, rs) ->
+      AccumT(AF()) { s: S ->
+        AF().run {
+          ls.runAccumT(MF(), s).orElse(rs.runAccumT(MF(), s))
+        }
+      }
+    }
+
+  override fun <A> empty(): Kind<AccumTPartialOf<S, F>, A> =
+    AccumT.liftF(AF(), AF().empty())
+}
+
+@extension
+interface AccumTApplicativeError<S, F, E> : ApplicativeError<AccumTPartialOf<S, F>, E>, AccumTApplicative<S, F> {
+  fun ME(): MonadError<F, E>
+
+  override fun MS(): Monoid<S>
+  override fun MF(): Monad<F> = ME()
+
+  override fun <A> raiseError(e: E): Kind<AccumTPartialOf<S, F>, A> =
+    AccumT.liftF(MF(), ME().raiseError(e))
+
+  override fun <A> Kind<AccumTPartialOf<S, F>, A>.handleErrorWith(f: (E) -> Kind<AccumTPartialOf<S, F>, A>): Kind<AccumTPartialOf<S, F>, A> =
+    this.fix().let { accumT ->
+      AccumT(MF()) { s: S ->
+        ME().run {
+          accumT.runAccumT(MF(), s).handleErrorWith { e ->
+            f(e).fix().runAccumT(MF(), s)
+          }
+        }
+      }
+    }
+}
+
+@extension
+interface AccumTMonadError<S, F, E> : MonadError<AccumTPartialOf<S, F>, E>, AccumTApplicativeError<S, F, E>, AccumTMonad<S, F> {
+  override fun MS(): Monoid<S>
+  override fun ME(): MonadError<F, E>
+  override fun MF(): Monad<F> = ME()
+}
+
+@extension
+interface AccumTMonadState<S, F> : MonadState<AccumTPartialOf<S, F>, S>, AccumTMonad<S, F> {
+  fun MSF(): MonadState<F, S>
+  override fun MF(): Monad<F> = MSF()
+  override fun MS(): Monoid<S>
+
+  override fun get(): Kind<AccumTPartialOf<S, F>, S> =
+    AccumT.liftF(MSF(), MSF().get())
+
+  override fun set(s: S): Kind<AccumTPartialOf<S, F>, Unit> =
+    AccumT.liftF(MSF(), MSF().set(s))
+}
+
+@extension
+interface AccumTMonadReader<S, F> : MonadReader<AccumTPartialOf<S, F>, S>, AccumTMonad<S, F> {
+  fun MR(): MonadReader<F, S>
+  override fun MF(): Monad<F> = MR()
+  override fun MS(): Monoid<S>
+
+  override fun ask(): Kind<AccumTPartialOf<S, F>, S> =
+    AccumT.liftF(MR(), MR().ask())
+
+  override fun <A> Kind<AccumTPartialOf<S, F>, A>.local(f: (S) -> S): Kind<AccumTPartialOf<S, F>, A> =
+    MR().run {
+      AccumT(fix().accumT.map { accumT ->
+        AndThen(accumT).andThen { it.local(f) }
+      })
+    }
+}
+
+@extension
+interface AccumTMonadWriter<S, F> : MonadWriter<AccumTPartialOf<S, F>, S>, AccumTMonad<S, F> {
```

Comment:
Same^^

Reply:
could not make the types work for the test. maybe you could have another look.

Reply:
got it now. this test fails:
```
java:test://arrow.mtl.AccumTTest.Monad Writer Laws: tell fusion
```

does this mean both instances are not lawful?